### DISCUSSION TITLE
NameError: name 'DBAPIConnection' is not defined

### DISCUSSION THREAD

**Discussion by usama-hameed** [sqlalchemy/sqlalchemy#10216] [Usage Questions]
I have installed llama_index and sqlalchemy is installed with it but on llama_index import I am getting this error, NameError: name 'DBAPIConnection' is not defined. I assume its is related to sqlalchemy.
Full Error:

[crain-api]     import sqlalchemy
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/__init__.py", line 13, in <module>
[crain-api]     from .engine import AdaptedConnection as AdaptedConnection
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/__init__.py", line 18, in <module>
[crain-api]     from . import events as events
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/events.py", line 19, in <module>
[crain-api]     from .base import Connection
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 30, in <module>
[crain-api]     from .interfaces import BindTyping
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/interfaces.py", line 36, in <module>
[crain-api]     from ..pool import Pool
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/__init__.py", line 20, in <module>
[crain-api]     from . import events
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/events.py", line 15, in <module>
[crain-api]     from .base import ConnectionPoolEntry
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 635, in <module>
[crain-api]     class _ConnectionRecord(ConnectionPoolEntry):
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 698, in _ConnectionRecord
[crain-api]     def connection(self) -> Optional[DBAPIConnection]:
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 141, in decorate
[crain-api]     return _decorate_with_warning(
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 396, in _decorate_with_warning
[crain-api]     decorated = warned(func)  # type: ignore
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 300, in decorate
[crain-api]     _exec_code_in_env(code, env, fn.__name__),
[crain-api]   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 334, in _exec_code_in_env
[crain-api]     exec(code, env)
[crain-api]   File "<string>", line 1, in <module>
[crain-api] NameError: name 'DBAPIConnection' is not defined

**Comment by CaselIT** [comment]
hi,

these error are almost always caused by issues in the user environment. I would try re-creating the environment from scratch

**Reply by usama-hameed** [reply]
Hi, Thanks for replying back,
Recreating the environment? can you plz elaborate a bit more, I didn't completely understate it.

**Reply by CaselIT** [reply]
python environment, usually created using virtualenv or similar. see https://docs.python.org/3/tutorial/venv.html

**Reply by usama-hameed** [reply]
Hello again, I am running application on kubernetes, all the packages are installed using requirements file in docker file. I have deleted all my images  from local docker registry and created all images from scratch again. But again I faced the same error, I am installing llama_index with llama_index sqlalchemy is installed automatically.

**Reply by CaselIT** [reply]
Seems to be an issue with llama_index or something else in your env.
I can't reproduce the error, so not sure how to help.

This was never reported before so it not something that usually happens by just importing sqlalchemy

**Comment by zzzeek** [comment]
it's not clear what's happening here because this is the deprecation decorator, it's invoking this code:

```
def connection(self) -> Optional[DBAPIConnection]:
    return target(fn, self)

```

`DBAPIConnection` is not available however this is an annotation that should not be invoked, because all files include `from __future__ import annotations`.

What Python interpreter / OS is this?  is this a special fork of Python ?

**Reply by usama-hameed** [reply]
I am using Ubuntu and this docker image python:3.8.5-buster

**Reply by CaselIT** [reply]
the following code seems to work without issues
```sh
podman run --rm python:3.8.5-buster bash -c 'pip install sqlalchemy; python -c "import sqlalchemy; print(sqlalchemy.__version__)"'
```
prints
```
2.0.19
```

**Comment by jonas-krug** [comment]


**Reply by CaselIT** [reply]
Do you have a list of packages and a python file that reproduces the error?

**Reply by CaselIT** [reply]
the link gives me "You need access" 

if you create a zip file you can upload it to a github message

**Reply by CaselIT** [reply]
the example code includes a lot of additional dependencies, like flask_bcrypt, flask_migrate

can you provide a complete requirement file?

also note that these are all 3rd party libraries, not developed by the sqlalchemy devs. If it's an easy thing to test we may look at what's causing issues with 3rd party libs, but I don't want to hunt down each and every dependency just to run the provided code

**Reply by jonas-krug** [reply]
I found the error. I thought the error is coming up with sqlalchemy because the errors caused in the sqlalchemy folder. But it was just the import from an 3rd-Party Libary, which was very well hidden


```
NameError: name 'DBAPIConnection' is not defined
``` 
is solved.

**Reply by CaselIT** [reply]
what's the library in question @jonas-krug

**Comment by Heanthor** [comment]
Hi, I'm seeing this same error after deploying on AWS Lambda with the Python 3.8 runtime and SQLAlchemy 2.0.20.

```
NameError: name 'DBAPIConnection' is not defined
Traceback (most recent call last):
  File "/var/lang/lib/python3.8/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/var/lang/lib/python3.8/imp.py", line 171, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 702, in _load
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/var/task/lambda_function.py", line 6, in <module>
    from sqlalchemy import create_engine
  File "/var/task/sqlalchemy/__init__.py", line 13, in <module>
    from .engine import AdaptedConnection as AdaptedConnection
  File "/var/task/sqlalchemy/engine/__init__.py", line 18, in <module>
    from . import events as events
  File "/var/task/sqlalchemy/engine/events.py", line 19, in <module>
    from .base import Connection
  File "/var/task/sqlalchemy/engine/base.py", line 30, in <module>
    from .interfaces import BindTyping
  File "/var/task/sqlalchemy/engine/interfaces.py", line 36, in <module>
    from ..pool import Pool
  File "/var/task/sqlalchemy/pool/__init__.py", line 20, in <module>
    from . import events
  File "/var/task/sqlalchemy/pool/events.py", line 15, in <module>
    from .base import ConnectionPoolEntry
  File "/var/task/sqlalchemy/pool/base.py", line 635, in <module>
    class _ConnectionRecord(ConnectionPoolEntry):
  File "/var/task/sqlalchemy/pool/base.py", line 698, in _ConnectionRecord
    def connection(self) -> Optional[DBAPIConnection]:
  File "/var/task/sqlalchemy/util/deprecations.py", line 141, in decorate
    return _decorate_with_warning(
  File "/var/task/sqlalchemy/util/deprecations.py", line 396, in _decorate_with_warning
    decorated = warned(func)  # type: ignore
  File "/var/task/sqlalchemy/util/langhelpers.py", line 300, in decorate
    _exec_code_in_env(code, env, fn.__name__),
  File "/var/task/sqlalchemy/util/langhelpers.py", line 334, in _exec_code_in_env
    exec(code, env)
  File "<string>", line 1, in <module>
```

Reading this stack trace I don't think any third party library is interacting with sqlalchemy at this point (the error is thrown by `from sqlalchemy import create_engine` at the top of the lambda function entrypoint file). The error is also thrown if I delete all other code in my lambda except for `from sqlalchemy import create_engine`. But, it's fine in a local Python 3.8 virtualenv.

Full requirements:
```
backports.zoneinfo==0.2.1
beautifulsoup4==4.12.2
boto3==1.28.29
botocore==1.31.29
certifi==2023.7.22
charset-normalizer==3.2.0
greenlet==2.0.2
idna==3.4
jmespath==1.0.1
psycopg==3.1.10
psycopg-binary==3.1.10
python-dateutil==2.8.2
requests==2.31.0
s3transfer==0.6.2
six==1.16.0
soupsieve==2.4.1
SQLAlchemy==2.0.20
typing_extensions==4.7.1
urllib3==1.26.16
```

**Reply by zzzeek** [reply]
OK does this environment have options for different Python versions?  Would be curious what python 3.9, 3.10 produce if possible.

I still cannot reproduce, and I have the exact same "imp" routine running like this:

```py
import imp

with open("test3.py") as f:
    imp.load_module("test", f, "test3.py", (".py", "r", imp.PY_SOURCE))
```

no problems running that on a python 3.8 interpreter here.

We can "fix" this one particular issue, however the problem is our CI would not catch future errors of this nature.   

can you try this patch and see if the rest of importing succeeds?

```diff
diff --git a/lib/sqlalchemy/pool/base.py b/lib/sqlalchemy/pool/base.py
index 7f542ae013..b4a0f2c8c8 100644
--- a/lib/sqlalchemy/pool/base.py
+++ b/lib/sqlalchemy/pool/base.py
@@ -695,7 +695,7 @@ class _ConnectionRecord(ConnectionPoolEntry):
         "The _ConnectionRecord.connection attribute is deprecated; "
         "please use 'driver_connection'",
     )
-    def connection(self) -> Optional[DBAPIConnection]:
+    def connection(self) -> "Optional[DBAPIConnection]":
         return self.dbapi_connection
 
     _soft_invalidate_time: float = 0

```

**Reply by CaselIT** [reply]
Don't we have future-amnotation on that file?
If so that change should have no effect, since the annotation is already a string 

The future import is also in langhelpers, so I'm not sure what's causing that error

**Reply by zzzeek** [reply]
The issue seems to be that future annotations mode is not taking effect within the exec() that the deprecation decorator is using.

I recall trying to understand when exec/eval use future annotations mode and it was kind of squirrelly, so the idea here is that whatever AWS lambda is doing with Python 3.8, those rules are somehow broken.

would be nice to know if AWS has their own build of python here and if it's broken, if there's a newer version of their python that's not broken, etc.

**Reply by CaselIT** [reply]
Maybe we can force the future-amnotation import by pre-pending it to what we are trying to exec?

**Reply by zzzeek** [reply]
Maybe ,.I had struggled w this previously.   Would be great if we could reproduce it but in the meantime seeing if the quote approach above fixes will work for now

**Reply by Heanthor** [reply]
I updated the Lambda environment to Python 3.10, and the error changed: 
```
START RequestId: 0fb438b0-1e21-4026-b437-89f6ea59baff Version: $LATEST
[ERROR] ModuleNotFoundError: No module named 'psycopg2'
Traceback (most recent call last):
  File "/var/task/lambda_function.py", line 23, in lambda_handler
    engine = create_engine(DB_STRING)
  File "<string>", line 2, in create_engine
  File "/var/task/sqlalchemy/util/deprecations.py", line 281, in warned
    return fn(*args, **kwargs)  # type: ignore[no-any-return]
  File "/var/task/sqlalchemy/engine/create.py", line 601, in create_engine
    dbapi = dbapi_meth(**dbapi_args)
  File "/var/task/sqlalchemy/dialects/postgresql/psycopg2.py", line 690, in import_dbapi
    import psycopg2END RequestId: 0fb438b0-1e21-4026-b437-89f6ea59baff
REPORT RequestId: 0fb438b0-1e21-4026-b437-89f6ea59baff	Duration: 1594.54 ms	Billed Duration: 1595 ms	Memory Size: 128 MB	Max Memory Used: 67 MB	Init Duration: 882.86 ms	

```
This one I was sort of expecting since I haven't compiled the psycopg binary for lambda yet (that seems like another huge mess but one I signed up for). So, it seems like that fixed this issue?

I will have some time tonight to try adding quotes to that import on 3.8, if you'd like. Since my local dev is on 3.8 I'd like to deploy there if possible.

**Reply by CaselIT** [reply]
@Heanthor that most likely unrelated. You forgot to inseall psycopg2 

Edit: but yes as you mentioned that seems after sqlalchemy is imported, so it should not have the previous error anymore 

Edit2: psycopg2 has binary wheels of you use psycopg2-binary as a package instead

**Reply by Heanthor** [reply]
Right, not asking for you to help me fix this one! Just letting you know the error changed in case there was anything else in this stack trace.

**Reply by zzzeek** [reply]
I took a look through the python issue tracker to see if anything looked like this but couldn't find it.  

what's the exact 3.8 version on AWS ?

**Reply by zzzeek** [reply]
it doesnt fail on 3.7 either, so it remains unknown what this actually is

**Comment by rvielmaa** [comment]
Guys
I have been working on my system, I think I am close to the information for you, and it is due to the existence of imports when wanting to find your session within Sqlalchemy, however, to know how to use this correctly you need recommendations from specific versions regarding and is of your system, I ask you to be able to understand what I am trying to explain to you in a more correct way and similar to what you do in your day. I recommend previous versions and that you know that they are stable to be able to correctly carry out your instance. and stable that can solve this error like the stable version, if you need to download it this is the correct command to do it. Doing this could be another possibility of a solution to what you may need to do:
![image](https://github.com/sqlalchemy/sqlalchemy/assets/132482329/825740b7-8272-47f0-9910-3a611de38544)

Stable Version
![image](https://github.com/sqlalchemy/sqlalchemy/assets/132482329/10986ee8-be30-47e6-ad56-4e47c3b735a0)
I hope to make helping

**Reply by rvielmaa** [reply]
Reminder
Continually review the import libraries you use to do it correctly.

**Comment by healer-21** [comment]
My error is the same as the owner's, and then I went to the official website to see that  the maintenance version of sqlalchemy is 1.4, changed his version to 1.4.52, the error was solved.
